### PR TITLE
fixes #17 - Handle duplicate email while user creation and updation

### LINKED ISSUE #17: Check if the email is unique in user create method
**Feature Description**
The create and update service methods should check if the user's email is unique. 

**Requirements**
**1.** The system should not allow duplicate emails.

### PR DESCRIPTION
**Pull request by foolmarshal**

### CODE REVIEW

**Review by inciotiq — COMMENTED** [review]
Hi @foolmarshal ,

I want to express my appreciation for your valuable contribution to our project! Unfortunately, our project is a commercial one, therefore I cannot accept your contribution in good conscience.

However, I want to emphasize that your effort is highly valued, and I'd like to offer some constructive feedback, just as I would with any other teammate. Your willingness to contribute is commendable and greatly appreciated.

Thank you once again for dedicating your time and expertise to open-source projects, and I wish you the best of luck in all your future endeavors!

File: src/main/java/com/iotiq/user/internal/UserRepository.java [hunk 1]
```diff
@@ -14,4 +15,8 @@ public interface UserRepository extends JpaRepository<User, UUID>, JpaSpecificat
 
     boolean existsByAccountInfoUsername(String username);
 
+    boolean existsByPersonalInfoEmail(String email);
+
+    @Query("SELECT COUNT(u) FROM User u WHERE u.personalInfo.email = :email AND u.id <> :id")
+    int countUsersWithPersonalInfoEmailAndNotId(UUID id, String email);
```
Comment (inciotiq) [inline src/main/java/com/iotiq/user/internal/UserRepository.java:21]: I think you could have also used the `exists` keyword here and refrain from writing a query.

File: src/main/java/com/iotiq/user/internal/UserService.java [hunk 2]
```diff
@@ -53,8 +54,20 @@ public boolean existByUserName(String username) {
         return userRepository.existsByAccountInfoUsername(username);
     }
 
+    @Transactional(readOnly = true)
+    public boolean existByEmail(String email) {
+        return userRepository.existsByPersonalInfoEmail(email);
+    }
+
+    public int countUsersWithEmailAndNotId(UUID id, String email) {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/user/internal/UserService.java:62]: This method calls a method with the same signature therefore seems unnecessary to me.

File: src/test/java/com/iotiq/user/UserApplicationTests.java [hunk 3]
```diff
@@ -111,6 +112,36 @@ void add() throws Exception {
 
     @Test
     @Order(2)
+    void addWithDuplicateEmail() throws Exception {
```
Comment (inciotiq) [inline src/test/java/com/iotiq/user/UserApplicationTests.java:115]: I appreciate you writing tests for your code :)

**Comment by foolmarshal** [comment]
Hi @inciotiq ,

Thanks for the feedback. It was such a delight to go through your codebase and understanding the code flow. I must say, the codebase follows a lot of good code practices and was really a learning experience for me. I will definitely learn from the constructive feedbacks you gave on the PR.

Thanks for taking out time for code review! 🙂

File: src/main/java/com/iotiq/user/internal/UserRepository.java (hunk 1 above)
Comment (foolmarshal) [reply src/main/java/com/iotiq/user/internal/UserRepository.java:21]: Noted!

File: src/main/java/com/iotiq/user/internal/UserService.java (hunk 2 above)
Comment (foolmarshal) [reply src/main/java/com/iotiq/user/internal/UserService.java:62]: Noted!

**Pull request closed without merging**